### PR TITLE
Use pre-commit

### PR DESCRIPTION
**Pull request by wsanchez**
Move listing from Tox config to `pre-commit` and use `pre-commit` to run them, along with additional checks.

This scans more files than before, so now our Python code in the documentation is checked, for example.

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+max-line-length = 80
+
+exclude =
+    .git,
+    .tox,
+    __pycache__,
+    dist
+
+# Codes: http://flake8.pycqa.org/en/latest/user/error-codes.html
+ignore =
```
Comment (graingert) [inline .flake8:18]:
```suggestion
extend-ignore =
```

File: .flake8 [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+# !!! BRING THE PAIN !!!
+select = A,B,B9,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z
+
+show-source = True
+doctests    = True
+
+max-line-length = 80
+
+exclude =
```
Comment (graingert) [inline .flake8:11]:
```suggestion
extend-exclude =
```

File: .flake8 [hunk 3]
```diff
@@ -0,0 +1,105 @@
+[flake8]
+
+# !!! BRING THE PAIN !!!
+select = A,B,B9,C,D,E,F,G,H,I,J,K,L,M,N,O,P,Q,R,S,T,U,V,W,X,Y,Z
```
Comment (graingert) [inline .flake8:4]: the default select is pretty good, this seems like a strange list

Comment (wsanchez) [reply .flake8:4]: I'm trying to enable all of the things, and not whatever things someone else picked.  Some plugins are not on by default…

In any case, this bit isn't new; it's been in the config for a while.

**Comment by wsanchez** [comment]
CI is not going to pass completely because we change the jobs that are run, so once this is approved, I'll need to change the `master` branch config in GitHub as part of merging.

File: src/klein/_form.py [hunk 4]
```diff
@@ -453,7 +452,8 @@ def defaultValidationFailureHandler(
 _requirerFunctionWithForm = Any
 _routeCallable = Any
 _routeDecorator = Callable[
-    [_routeCallable, DefaultNamedArg(Any, "__session__")], _routeCallable
+    [_routeCallable, DefaultNamedArg(Any, "__session__")],  # noqa: F821
+    _routeCallable,
```
Comment (wsanchez) [inline src/klein/_form.py:456]: These type aliases should go away.  See https://github.com/twisted/klein/issues/414

File: MANIFEST.in [hunk 5]
```diff
@@ -1,5 +1,7 @@
 include .codecov.yml
 include .coveragerc
+include .flake8
+include .pre-commit-config.yaml
```
Comment (graingert) [inline MANIFEST.in:4]:
```suggestion
include .pre-commit-config.yaml
```

Comment (graingert) [reply MANIFEST.in:4]: pre-commit doesn't work without the .git directory so there's no use having it in the sdist

**Comment by wsanchez** [comment]
OK, going once: anyone opposed to using `pre-commit` here?

File: docs/_extensions/apilinks.py [hunk 6]
```diff
@@ -43,8 +53,11 @@ def make_api_link(name, rawtext, text, lineno, inliner,
...
 
+
 def setup(app):
-    app.add_config_value('apilinks_base_url',
-                         'http://twistedmatrix.com/documents/current/api/',
-                         'env')
-    app.add_role('api', make_api_link)
+    app.add_config_value(
+        "apilinks_base_url",
+        "http://twistedmatrix.com/documents/current/api/",
```
Comment (twm) [inline docs/_extensions/apilinks.py:60]:
```suggestion
        "https://twistedmatrix.com/documents/current/api/",
```

File: docs/conf.py [hunk 7]
```diff
@@ -1,45 +1,58 @@
...
-templates_path = ['_templates']
-source_suffix = '.rst'
-master_doc = 'index'
-project = 'Klein'
-copyright = '2011-2019, Twisted Matrix Labs'
+templates_path = ["_templates"]
+source_suffix = ".rst"
+master_doc = "index"
+project = "Klein"
+copyright = "2011-2019, Twisted Matrix Labs"
```
Comment (twm) [inline docs/conf.py:15]:
```suggestion
copyright = "2011-2021, Twisted Matrix Labs"
```

File: docs/introduction/1-gettingstarted.rst [hunk 8]
```diff
@@ -62,9 +62,9 @@ Add more decorated functions to add more routes to your Klein applications.
 Variable Routes
```
Comment (twm) [inline docs/introduction/1-gettingstarted.rst:62]: This file contains references to `codeexamples/template.py` with ``:lines:``. Those line numbers need an update (or switch them to `:pyobject:` if that works with the decorators.

**Review by twm — APPROVED** [review]
IMO, if you do the work, you get to pick the tools. And you're certainly running this show. :)

I did a quick skim of what I assume are mostly changes made by Black. Apart from some documentation updates this looks good to me.

File: docs/_extensions/apilinks.py (hunk 6 above)
Comment (wsanchez) [reply docs/_extensions/apilinks.py:60]: Fixed in a bunch of places

File: docs/conf.py (hunk 7 above)
Comment (wsanchez) [reply docs/conf.py]: Fixed in a bunch of places

File: docs/introduction/1-gettingstarted.rst (hunk 8 above)
Comment (wsanchez) [reply docs/introduction/1-gettingstarted.rst:62]: Fixed the line numbers, which were already wrong anyway

**Pull request merged**